text/0015-status-update.md

```diff
@@ -0,0 +1,397 @@
+  this dramatically constrains our solution space, and will lead to a design
+  that is seriously compromised in the general case.
+- I would prefer to focus on the general case, and then try to figure out a more
+  holistic approach to the subsetting problem. In practical terms, this means a
+  general approach that does more transient allocation for subtables etc, and
+  then a subsetting approach that (for instance) uses a long-running process to
+  handle multiple subsetting requests, and reuses memory and other state between
+  them.
+
+## next steps
+
+### a design for mutability
+
+My main concern right now is figuring out a model for compilation & mutability.
+For compilation, I'm currently imagining generating builder types, but I haven't
+dug into this yet.
+
+Offset resolution and rewriting will also be an important consideration. A
+compiler author should be able to use these tools to do things like deduplicate
+identical subtables, or to reorder tables as needed, which means we do need a
+mechanism for rewriting offsets as needed. There are various options here; for
+instance we could track offsets by ids, and then keep track of the location of
+the offsets as we write tables into a shared buffer, rewriting them later as
+needed. This will need more thought, experimentation, and discussion.
+
+### evaluating current performance
+
+It may also be useful to get some concrete numbers on the efficiency of the
+current work-to-date. I have spent a bit of time looking at disassembly and am
+generally confident that we should be able to achieve good results, but I
+haven't *demonstrated* this.
+
+I'm not sure this is strictly necessary at this point, but it might be helpful;
+if performance is actually terrible for some unforeseen reason, we might decide
+that our approach needs to be fundamentally rethought.
+
+I've also resisted this because I'm not sure of exactly what we should measure.
+In any case, it is something we will want to do at some point.
+
+### further refinements
+
+Finally, there are various additional features and functionality that it might
+be worth exploring, such as improving type information related to offsets (for
+instance encoding in the type of the offset what sort of table it should
+produce, or what table it is indexed into.) In addition, I am *very curious*
+about looking into migrating from doing codegen in a proc-macro to doing codegen
+as a separate build script that outputs normal rust files that are then included
+and used like any other. This has several advantages: it should reduce
+compilation time, but it should also produce more readable code: currently we
+generate lots of methods and type definitions that do not actually exist in any
+of our source files. This makes it harder for users to understand what is going
+on, and it also confuses tools like IDEs, which often cannot provide
+autocomplete for these generated methods.
```

Comment:
How long do you think an exploration of separate build script would take?

(how long would be good to address for all the potential future work)

Reply:
There are some big unknowns here for me: basically as long as I can reuse all of the proc-macro parsing & code-generation code, it shouldn't be that hard. So I would say it should take me a day to figure out if that's possible, and then if it is it would take another week to move over in this direction. The major unknowns are basically: I want my final result to be a complete crate (or *perhaps* a module) and this will require me to come up with a way to stitch together the generated code with small bits of hand-written stuff, so I'd need to figure out how I want to do that: either having the manual code inlined, or else having a sort of simple templating system.

Reply:
I think perhaps it's worth noting the estimates in the doc? - even with big error bars A is 1-2 weeks and B is 1-2 quarters is informative.